### PR TITLE
Implemented working Name tag

### PR DESCRIPTION
**Pull request by IvanCraft623**
## Introduction

Implement the nametags and their function (currently only obtainable through commands because there are no anvils).

## Follow-up
- https://github.com/pmmp/Language/pull/134

## Tests

https://youtu.be/DCpjGbXHEh0
(The interactive tag for fishing rod was implemented overriding the item on a plugin for testing the api)

### CODE REVIEW

**Comment by xerenahmed** [comment]
I think an event must be written for this feature.

**Comment by IvanCraft623** [comment]
> I think an event must be written for this feature.

There is PlayerEntityInteractEvent

File: src/entity/Living.php [hunk 1]
```diff
@@ -352,6 +353,19 @@ protected function onHitGround() : ?float{
 		return $newVerticalVelocity;
 	}
 
+	public function onInteract(Player $player, Vector3 $clickPos) : bool{
+		$item = $player->getInventory()->getItemInHand();
+		if($item->getTypeId() === ItemTypeIds::NAME_TAG && !$this instanceof Human && $item->hasCustomName()){
```
Comment (dktapps) [inline src/entity/Living.php]: `$this instanceof` defeats the purpose of OOP. Also, I'm not sure this logic belongs in `Living`.

Comment (IvanCraft623) [reply src/entity/Living.php]: > I'm not sure this logic belongs in `Living`

According to my tests and the wiki, all Living entities with the exception of EnderDragon and Human can be named with the nametag.

Comment (IvanCraft623) [reply src/entity/Living.php]: > `$this instanceof` defeats the purpose of OOP

Solved.

**Review by xRodzMC — DISMISSED** [review]

**Comment by xRodzMC** [comment]
Looks good, well done

**Comment by dktapps** [comment]
The PR looks functionally OK.

The thing that's giving me pause is that the mode of operation here is inconsistent with every other method of item interaction. For example, weapon damage when attacking an entity is not handled by the entity itself; it's processed in `Item->onAttackEntity()`. We also already have API methods like `Item->onInteractBlock()`.

**Comment by IvanCraft623** [comment]
> The PR looks functionally OK.
> 
> The thing that's giving me pause is that the mode of operation here is inconsistent with every other method of item interaction. For example, weapon damage when attacking an entity is not handled by the entity itself; it's processed in `Item->onAttackEntity()`. We also already have API methods like `Item->onInteractBlock()`.

I agree with adding a method like that, but for this specific case I am more inclined towards the idea of adding an interface for renameable entities because they have a property: `EntityMetadataProperties::INTERACTIVE_TAG`.

**Comment by IvanCraft623** [comment]
The following changes have been made:
 - Logic for renaming has been moved to `NameableTrait`, renameable entities should use it and implement `Nameable`
 - Implemented API for Interactive Tags (interaction button displayed above the hotbar).
 - Interactive tag is now displayed when looking at the renameable entity.

The above changes require [this](https://github.com/pmmp/Language/pull/134) to be merged

**Comment by IvanCraft623** [comment]
This PR has caught up with the latest changes and now uses the new `Item::onInteractEntity()` method.

File: src/entity/NameableTrait.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	/**
+	 * @see Entity::getInteractiveTag()
+	 */
+	public function getInteractiveTag(Player $player, Item $item) : ?string{
+		if($item->getTypeId() === ItemTypeIds::NAME_TAG && $item->hasCustomName()){
+			return KnownTranslationKeys::ACTION_INTERACT_NAME;
+		}
+		return parent::getInteractiveTag($player, $item);
+	}
```
Comment (dktapps) [inline src/entity/NameableTrait.php]: This seems redundant. It can be implemented in Living directly.

File: src/entity/Nameable.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+declare(strict_types=1);
+
+namespace pocketmine\entity;
+
+use pocketmine\item\Item;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+
+interface Nameable{
```
Comment (dktapps) [inline src/entity/Nameable.php]: This interface seems redundant.

File: src/player/Player.php [hunk 4]
```diff
@@ -607,6 +611,36 @@ public function setDisplayName(string $name) : void{
...
+	 */
+	public function setLookingEntity(?Entity $looking) : void{
+		if($looking === null){
+			$this->lookingId = null;
+		}elseif($looking->closed){
+			throw new \InvalidArgumentException("Supplied looking entity is garbage and cannot be used");
+		}else{
+			$this->lookingId = $looking->getId();
+		}
+	}
```
Comment (dktapps) [inline src/player/Player.php:635]: This entire block looks redundant.

File: src/player/Player.php [hunk 5]
```diff
@@ -1852,6 +1887,23 @@ public function interactEntity(Entity $entity, Vector3 $clickPos) : bool{
...
+		$item = $this->inventory->getItemInHand();
+		$interactiveTag = ($looking !== null ? $looking->getInteractiveTag($this, $item) : null) ?? $item->getInteractiveTag();
+		if($interactiveTag !== $this->interactiveTag){
+			$this->interactiveTag = $interactiveTag;
+			$this->networkPropertiesDirty = true;
+			$this->sendData([$this]);
+			return true;
+		}
+		return false;
+	}
```
Comment (dktapps) [inline src/player/Player.php]: This also looks redundant.

**Review by dktapps — DISMISSED** [review]
The interactive tags stuff looks like a really overengineered mess and can be greatly simplified.

File: src/player/Player.php (hunk 5 above)
Comment (IvanCraft623) [reply src/player/Player.php]: What is redundant?

**Comment by IvanCraft623** [comment]
I do not understand what is redundant
`Nameable` and `NameableTrait` exists because not all `Living` entities are nameable, some examples are:
- `Human`
- `EnderDragon`
and there are also non-`Living` entities that are nameable:
- `ArmorStand`

**Comment by jasonw4331** [comment]
I believe the look detection is what is being considered as redundant here. The nametag can be implemented from when it gets used and shouldn't depend on whether or not an entity is being looked at.

**Comment by dktapps** [comment]
The look detection isn't redundant in itself, I just think that the implementation of it is vastly overcomplicated.

**Pull request merged**